### PR TITLE
Support for kprobe:[module]:[function]

### PR DESCRIPTION
**Pull request by matthew-olson-intel**
Particularly useful in situations where you need BTF symbol information for functions in a kernel module. Passes all (previously existing) tests (on my system).

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by matthew-olson-intel** [comment]
I'm currently working on fixing/adding some unit tests; while the current state *works* in terms of providing a kernel module's BTF symbols in a kprobe, it doesn't pass the tests "properly."

**Comment by danobi** [comment]
Sounds good. Obviously no rush

File: tests/parser.cpp
Comment (matthew-olson-intel) [inline tests/parser.cpp:1]: I was a little iffy about changing this, but my reasoning is that this test originally failed because a dangling colon indicates that there's a second (empty) argument to that probe, so it should fail because a `kprobe` should take one argument. With this change, though, the test now ensures that the parser fails on adding a _third_ (also empty) argument.

File: docs/reference_guide.md
Comment (jordalgo) [inline docs/reference_guide.md:1123]: What is kvm.bt?

File: src/ast/attachpoint_parser.cpp
Comment (jordalgo) [inline src/ast/attachpoint_parser.cpp:328]: nit: maybe save off a local var e.g. `auto parts_size = parts_.size()`

File: man/adoc/bpftrace.adoc
Comment (jordalgo) [inline man/adoc/bpftrace.adoc:2394]: nit: maybe just spell this out mod -> module

File: tests/runtime/probe
Comment (jordalgo) [inline tests/runtime/probe:126]: Is it worth adding a runtime test for a kernel module that's either not loaded and/or doesn't exist?

**Comment by jordalgo** [comment]
Few small comments but otherwise LGTM!

File: docs/reference_guide.md
Comment (matthew-olson-intel) [reply docs/reference_guide.md:1123]: Oh, it's just a made-up name of a file: similar to how the other multi-line examples are in the Reference Guide.

File: src/ast/attachpoint_parser.cpp
Comment (matthew-olson-intel) [reply src/ast/attachpoint_parser.cpp:328]: Yeah, that does look nicer.

File: man/adoc/bpftrace.adoc
Comment (matthew-olson-intel) [reply man/adoc/bpftrace.adoc:2394]: I think that would look nicer, but I was copying what the `kfunc` documentation does above -- I'll change both.

File: tests/runtime/probe
Comment (matthew-olson-intel) [reply tests/runtime/probe:126]: Good idea; we need to make sure that not only does the PR *work* the same way as without a module being specified, but that it also *fails* in the same way.

Comment (matthew-olson-intel) [reply tests/runtime/probe:126]: ...Adding this revealed a bit of a problem. If I specify a kernel module, attaching appears to "succeed." That is, the `bpf_attach_kprobe` call (passing in `module:function` as an argument) succeeds in creating a perf fd. I'll continue discussing on the affected line.

File: src/attached_probe.cpp
Comment (matthew-olson-intel) [inline src/attached_probe.cpp:975]: The previous PR (that I based mine on) used [this](https://www.kernel.org/doc/Documentation/trace/kprobetrace.rst) as justification for this working, but that link doesn't refer to libbpf. The real question is whether the underlying `perf_event_open` call in [libbpf](https://github.com/iovisor/bcc/blob/d9645f8253d1d8aed0d43503e70dacfb88cb1c72/src/cc/libbpf.c#L1128) is able to take a `module:function` (and properly fail if either doesn't exist).

File: docs/reference_guide.md
Comment (jordalgo) [reply docs/reference_guide.md:1123]: Ha, I see them now. I wonder why those exist when they don't point to actual bpftrace scripts in the tools folder. Seem a bit weird to me - maybe I'll make another PR to remove them all 😃

File: src/attached_probe.cpp
Comment (jordalgo) [reply src/attached_probe.cpp:975]: So what happens if the module is some junk string? Is there an eventual failure somewhere else?

Comment (matthew-olson-intel) [reply src/attached_probe.cpp:975]: All right. After digging around in `libbpf` a little deeper, that name just gets passed to `perf_event_open` directly (e.g. it doesn't split the module out from the function name). So, we need to compare bpftrace's behavior versus directly using the kprobe sysfs.

With this PR, `bpftrace` exhibits the following behavior:

```
# Good module, good function
./src/bpftrace -e 'kprobe:kvm:x86_emulate_insn {1}'
[SUCCESS]

# Good module, bad function
./src/bpftrace -e 'kprobe:kvm:nonsense {1}'
[ERROR]

# Bad module, good function
./src/bpftrace -e 'kprobe:nonsense:x86_emulate_insn {1}'
[SUCCESS]

# Bad module, bad function
./src/bpftrace -e 'kprobe:nonsense:nonsense {1}'
[WARNING]
```

Particularly the configuration that shows a bad module *and* a bad function
made me think that this isn't getting passed along to `perf` correctly;
perhaps `bpf_attach_kprobe` doesn't properly handle `module:function`.

*But* if we use the sysfs interface directly, we get the following behavior:

```
# echo 'p:gg kvm:x86_emulate_insn' > /sys/kernel/tracing/kprobe_events
# echo 'p:gb kvm:nonsense' > /sys/kernel/tracing/kprobe_events
bash: echo: write error: No such file or directory
# echo 'p:bg nonsense:x86_emulate_insn' > /sys/kernel/tracing/kprobe_events
# echo 'p:bb nonsense:nonsense' > /sys/kernel/tracing/kprobe_events
```

... So the only configuration that fails is `gb`, when the module is good
but the function is bad, which exactly matches the new behavior of `bpftrace`.

I think it should be good, but let me know if you think we should do additional module-checking.

Comment (jordalgo) [reply src/attached_probe.cpp:975]: How complicated is it to check if the module is bad? Even though this mimics the sysfs behavior, I think it would be nice to give users some better messaging. Also should this succeed if the module is bad but the function is good? My gut reaction is no.

File: src/bpftrace.cpp
Comment (viktormalik) [inline src/bpftrace.cpp:329]: nit: merge into the previous case?

File: src/driver.cpp
Comment (viktormalik) [inline src/driver.cpp:93]: Should we check that `ap->target` is non-empty?

Also please update the top comment for this function.

**Review by viktormalik — COMMENTED** [review]
Thanks for picking up this work! Just a couple of comments.

Also what about probe listing? At the moment, running `bpftace -l kprobe:kvm:*` lists all kprobes while I'd expect it to list those from the `kvm` module only. I'm ok if we add that in a separate PR, just wanted to make sure that it gets added eventually.

File: src/attached_probe.cpp
Comment (viktormalik) [reply src/attached_probe.cpp:975]: Agreed with @jordalgo here. Even though ftrace silently succeeds with probe attachment to an invalid module, it'd be nice if we could print a warning that the module is not loaded.

File: src/bpftrace.cpp
Comment (matthew-olson-intel) [reply src/bpftrace.cpp:329]: I only unmerged the conditions for aesthetic reasons; since `kprobe` and `kretprobe` require the additional `attach_point->target` check, the indentation looks off to me. Will change though.

File: docs/reference_guide.md
Comment (matthew-olson-intel) [reply docs/reference_guide.md:1121]: Ach, forgot to mention that use-case.

**Comment by matthew-olson-intel** [comment]
> Also what about probe listing? At the moment, running `bpftace -l kprobe:kvm:*` lists all kprobes while I'd expect it to list those from the `kvm` module only. I'm ok if we add that in a separate PR, just wanted to make sure that it gets added eventually.

Well, it definitely shouldn't list _all_ kprobes, but this might be a bit tricky; e.g.:
```
sudo ./src/bpftrace -vl kfunc:kvm:*
```
returns nothing on my system. Same for other kernel modules (e.g. `i915`, which definitely has kfuncs).

File: src/attached_probe.cpp
Comment (matthew-olson-intel) [reply src/attached_probe.cpp:975]: Digging a bit further, `kfunc` seems to rely on the availability of BTF for the given module/function, so fails in 3/4 cases (as it should) based on those grounds:
```
ERROR: No BTF found for kvm:nonsense
ERROR: No BTF found for nonsense:x86_emulate_insn
ERROR: No BTF found for nonsense:nonsense
```

Comment (matthew-olson-intel) [reply src/attached_probe.cpp:975]: It looks like the way forward might be to simply parse `/proc/modules` (for example, in a function in `src/utils.cpp`), and call that before attaching a kprobe that specifies a module. Or we could just throw a warning based on that parsing.

That seems to be how the `lsmod` implementation in busybox does it:
https://github.com/brgl/busybox/blob/abbf17abccbf832365d9acf1c280369ba7d5f8b2/modutils/lsmod.c#L78

And kmod, too:
https://github.com/kmod-project/kmod/blob/48b0c75f79f86140314b9656d3d855a515e18a73/libkmod/libkmod-module.c#L1727

Comment (jordalgo) [reply src/attached_probe.cpp:975]: Yah, I was pretty much going to suggest doing whatever `lsmod` does. So that works for me.

Comment (matthew-olson-intel) [reply src/attached_probe.cpp:975]: Great; does that solution (a function to parse `/proc/modules`) sound good to @viktormalik too?

**Comment by viktormalik** [comment]
> Well, it definitely shouldn't list _all_ kprobes, but this might be a bit tricky; e.g.:
> 
> ```
> sudo ./src/bpftrace -vl kfunc:kvm:*
> ```
> 
> returns nothing on my system. Same for other kernel modules (e.g. `i915`, which definitely has kfuncs).

That is interesting, I can reproduce that but I'm pretty sure that it worked a while ago. Let me have a look.

File: src/attached_probe.cpp
Comment (matthew-olson-intel) [reply src/attached_probe.cpp:975]: @jordalgo @viktormalik My opinion is that we should mimic the `kfunc` behavior; if the module doesn't exist, bpftrace should immediately throw an error (similar to how failing in libbpf might behave).  Thoughts?

**Comment by matthew-olson-intel** [comment]
Fixing wildcards; they're completely broken with this PR.

By the way, in the case where we want to attach to a function that exists in two different modules, there aren't any examples of that on my system. I just printed out the module and function (separated by a space) from `get_symbols_from_traceable_funcs` and ran it through `sort | uniq -d -f 1`. We could obviously write a kernel module that causes that, but that'd be a pain to set up in CI.

File: src/attached_probe.cpp
Comment (viktormalik) [reply src/attached_probe.cpp:999]: Agreed, parsing `/proc/modules` makes sense and throwing an error if we cannot attach does, too. For wildcarded module attachpoints, we could just print a warning.

**Comment by viktormalik** [comment]
> Fixing wildcards; they're completely broken with this PR.

I fixed the problem with kfunc wildcards (#2914) and also refactored probe matching for kprobe/kfunc a bit (#2915). With those 2 PRs merged, adding support for `kprobe:module:*` listing should be easier.
 
> By the way, in the case where we want to attach to a function that exists in two different modules, there aren't any examples of that on my system. I just printed out the module and function (separated by a space) from `get_symbols_from_traceable_funcs` and ran it through `sort | uniq -d -f 1`. We could obviously write a kernel module that causes that, but that'd be a pain to set up in CI.

When I do `cat /sys/kernel/tracing/available_filter_functions | sort | uniq | cut -d" " -f1 | uniq -d`, I can see quite a lot of functions. For instance:
```
# cat /sys/kernel/tracing/available_filter_functions | grep cleanup_entry
cleanup_entry [ip_tables]
cleanup_entry [ip6_tables]
```

But I agree that adding that to CI may be quite painful (which is why I didn't do that when adding support for `kfunc:module:function` syntax).

**Comment by matthew-olson-intel** [comment]
> I fixed the problem with kfunc wildcards (#2914) and also refactored probe matching for kprobe/kfunc a bit (#2915). With those 2 PRs merged, adding support for `kprobe:module:*` listing should be easier.

Sure, I'll wait until those are merged and finish up wildcard support. Working for `-l`, but not for bpftrace programs yet (but your PR should fix it).

Then I'll write some more runtime tests, since we need to additionally test listing and probing modules and functions with wildcards:

```
-l kprobe:kvm:x86_emulate_ins*
-e 'kprobe:kvm:*'
-e 'kprobe:vmlinu*:vfs_read'
...
```

**Comment by matthew-olson-intel** [comment]
All the above problems are fixed, now, including:
1. Listing kprobes.
2. Wildcards (for modules and functions).
3. More unit tests.
4. Checking `/proc/modules` to throw errors/warnings if the attachment is going to fail.
5. Avoidance of the `kprobe_multi` codepaths which were previously causing attachment errors.

File: src/attached_probe.cpp
Comment (jordalgo) [inline src/attached_probe.cpp:966]: nit
```suggestion
  // Also log a warning or error if the module doesn't exist,
```

File: src/attached_probe.cpp
Comment (jordalgo) [inline src/attached_probe.cpp:970]:
```suggestion
  std::string& modname = probe_.path;
```

File: src/attached_probe.cpp
Comment (jordalgo) [inline src/attached_probe.cpp:978]: If the module isn't loaded - how do we know it doesn't exist? Rather why doesn't this mirror the error message below?

File: src/utils.cpp
Comment (jordalgo) [inline src/utils.cpp:709]: Maybe instead of the split we can just compare against the substring

File: src/utils.h
Comment (jordalgo) [inline src/utils.h:201]:
```suggestion
bool is_module_loaded(const std::string &module);
```

File: tests/runtime/btf
Comment (jordalgo) [inline tests/runtime/btf:99]: Is it possible to write a BEFORE program to load the kvm module if it's not already loaded?

**Review by jordalgo — CHANGES_REQUESTED** [review]
Some minor comments but otherwise looks pretty good.

File: src/attached_probe.cpp
Comment (viktormalik) [inline src/attached_probe.cpp:971]: Any reason not to do this?
```suggestion
  if ((modname.length() > 0) && modname != "vmlinux"))
```

File: src/bpftrace.cpp
Comment (viktormalik) [inline src/bpftrace.cpp:316]: Also update the below comment please.

File: src/attached_probe.cpp
Comment (viktormalik) [inline src/attached_probe.cpp:765]: Should we explain the reason for avoiding `kprobe_multi` in this case?

File: src/driver.cpp
Comment (viktormalik) [inline src/driver.cpp:95]: Future work idea: we could eventually drop the last part of the condition and parse BTF also for module functions specified in the form `kprobe:function` (as we do for kfuncs).

File: src/utils.cpp
Comment (viktormalik) [inline src/utils.cpp:698]: Same as above:
```suggestion
  if (module == "vmlinux")
```

**Review by viktormalik — COMMENTED** [review]
Couple of additional comments/suggestions. Otherwise looks nice!

File: src/driver.cpp
Comment (matthew-olson-intel) [reply src/driver.cpp:95]: Yeah, that's a good idea too. Probably wouldn't be too tough, either.

File: src/attached_probe.cpp
Comment (matthew-olson-intel) [reply src/attached_probe.cpp:978]: It should definitely mirror the below error message.

File: tests/runtime/btf
Comment (matthew-olson-intel) [reply tests/runtime/btf:99]: Good idea. I'll do that in another PR, since I'd rather not sneak it into this one; there may be a good reason that it's not already done.

**Comment by matthew-olson-intel** [comment]
Sorry about the spam. That's what I get for clicking the "Commit Changes" button in GitHub on each of the suggestions instead of properly committing them; they bypassed my testing and my pre-commit formatting hook.

File: src/attached_probe.cpp
Comment (viktormalik) [inline src/attached_probe.cpp:763]: I don't like the wording too much, what about
```suggestion
      // We want to avoid kprobe_multi when a module is specified because the BPF_TRACE_KPROBE_MULTI
      // link type does not currently support the `module:function` syntax.
```

File: src/attached_probe.cpp
Comment (viktormalik) [inline src/attached_probe.cpp:981]: Could we prepare the message into a string and then send it either to `LOG(WARNING)` or `LOG(ERROR)`? That way, when we want to change the message, we just do it in one place.

File: src/utils.cpp
Comment (viktormalik) [inline src/utils.cpp:709]: Nit: this should be slightly faster as it requires just one pass through the string:
```suggestion
    if (line.compare(0, module.size() + 1, module + " ") == 0)
```

File: src/utils.h
Comment (viktormalik) [reply src/utils.h:201]: Still applies, right?

Comment (matthew-olson-intel) [reply src/utils.h:201]: Yep, for sure. Got lost in the formatting fixup that I did after clicking the "Commit suggestion" button. Sorry about that.

**Comment by matthew-olson-intel** [comment]
@viktormalik Applied all suggestions. Thank you again for the review!

**Review by viktormalik — APPROVED** [review]
Thanks for the work!

Let's wait for at least one more approval and we'll squash the PR afterwards.

File: src/probe_matcher.cpp
Comment (jordalgo) [inline src/probe_matcher.cpp:121]:
```suggestion
      symbol_stream = get_symbols_from_traceable_funcs(!target.empty());
```

File: src/utils.cpp
Comment (jordalgo) [inline src/utils.cpp:701]: Even though I hate this kernel style convention. I think we don't use brackets for single statements under conditionals
```suggestion
  if (module == "vmlinux")
    return true;
```

File: tests/bpftrace.cpp
Comment (jordalgo) [inline tests/bpftrace.cpp:29]:
```suggestion
  if (!target.empty())
    return "kprobe:" + target + ":" + attach_point + str;
```

**Review by jordalgo — APPROVED** [review]
Thanks for the fast follow-ups! Looks great. Just a few more minor style nits but I'll let the maintainers decide if you want to skip them.

File: src/utils.cpp
Comment (matthew-olson-intel) [reply src/utils.cpp:701]: Ah, OK, I've got no issue with changing this. Didn't know that `clang-format` didn't support this (but a quick Google reveals it's a relatively new feature).

Comment (jordalgo) [reply src/utils.cpp:701]: @matthew-olson-intel Actually you can ignore my comment. Seems like bpftrace doesn't require the removing of these brackets - at least according to @viktormalik

File: src/probe_matcher.cpp
Comment (matthew-olson-intel) [reply src/probe_matcher.cpp:121]: I think this makes the intention of the code slightly more difficult to read, even if it's a few more lines, so I'd prefer the original.

File: src/utils.cpp
Comment (matthew-olson-intel) [reply src/utils.cpp:701]: All right -- either way!

**Comment by viktormalik** [comment]
Merged, thanks for the work!

**Pull request merged**
